File: src/nodes/data/u32.rs
use apica_common::bytecodes::ApicaSpecificationBytecode;
use crate::utils::position::Position;

pub struct NodeDataU32 {
    position: Position,
    bytecode: ApicaSpecificationBytecode,
    value: u32,
}

impl NodeDataU32 {
    pub fn init(position: Position, bytecode: ApicaSpecificationBytecode, value: u32) -> NodeDataU32 {
        NodeDataU32 { position, bytecode, value }
    }

    pub fn get_position(&self) -> &Position {
        &self.position
    }

    pub fn show(&self, indent: &mut String, end: char) {
        print!("{indent}NodeDataU32(data: {:?}, value: {}){end}", self.bytecode, self.value);
    }

    pub fn get_bytecode(&self) -> ApicaSpecificationBytecode {
        self.bytecode
    }

    pub fn get_value(&self) -> u32 {
        self.value
    }
}

File: src/nodes/data/u32.cpp
#include "nodes/data/u32.hpp"
#include "core/emitter.hpp"
#include <iostream>

using namespace nodes;

NodeDataU32::NodeDataU32(const utils::Position &position, common::bytecodes::ApicaSpecificationBytecode spec_bytecode, uint32_t value)
    : NodeData(position, spec_bytecode), value(value) {

}

void NodeDataU32::show(std::string &indent, char end) const {
    std::cout << indent << "NodeDataU32(spec: " << this->spec_bytecode << ", val: " << this->value << ')' << end;
}

NodeKind NodeDataU32::getKind() const {
    return NodeKind::DataU32;
}

void NodeDataU32::emit(core::Emitter &emitter) const {
    emitter.writeU64(this->spec_bytecode);
    emitter.writeU32(this->value);
}

uint32_t NodeDataU32::getSpecificationValue() const {
    return this->value;
}